### PR TITLE
Provide functionality similar to .NET Web API

### PR DESCRIPTION
**Pull request by dcherman**
## Description

These changes introduce `HttpResponseMessage` and `IHttpActionResult` APIs as a supported way of returning responses from controllers objects.  We then handle these return types on our end in order to decouple controller methods from directly needing to know about Response objects.  By doing so, we enable far easier unit testing of controller methods, similar to what you can accomplish in the .NET world.

This also introduces a variety of helper methods for common response types on our BaseHttpController.  Notably excluded from this initial deliverable are MediaTypeFormatters and Content Negotiation; I've left that for a follow-up merge request if needed since I wanted to start getting feedback on this proposal sooner rather than later.

Documentation is still a TODO.  Would like to get some feedback on this approach to make sure it's the route we want to go down before spending time on the documentation.

## Motivation and Context

https://github.com/inversify/InversifyJS/issues/779

## How Has This Been Tested?

Lots of unit tests and/or integration tests.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/results/ResponseMessageResult.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { HttpResponseMessage } from "../httpResponseMessage";
+import { CREATED } from "http-status-codes";
```
Comment (dcherman) [inline src/results/ResponseMessageResult.ts]: Well this is wrong.  I obviously missed this when I was making some templates, but I'm pretty surprised linting didn't catch this.

Want me to look into fixing the lint rules, or just fix the imports?

File: src/httpResponseMessage.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { OutgoingHttpHeaders } from "http";
+
+export class HttpResponseMessage<T> {
```
Comment (dcherman) [inline src/httpResponseMessage.ts]: Looking at the .NET implementation, they don't use generics for this, but rather provide an `HttpContent` abstract class that gets subclassed to specific typed like `StringContent`, `StreamContent`, MultipartContent`, etc.

May be worth a discussion even though it is more complex than simply using generics here which is the quick/easy route.

**Comment by AltekkeE** [comment]
@dcherman Sorry that it has taken so long to get back to you. This is looking great. Could you please update the documentation?

Thanks

**Comment by dcherman** [comment]
@AltekkeE Sure, glad to see that this is looking good.  Might be a couple days before I can address it, but I'll get it done.  Also still considering some of my inline comments about the usage of generics with `HttpResponseMessage` versus how it works in .NET with `HttpContent` and the various subclasses.

**Comment by dcherman** [comment]
@dcavanagh Sorry for the delay here, I've added some documentation for the new functionality.  I also refactored the HttpResponseMessage to no longer use generics, but expect that an `HttpContent` object be provided that'll contain the content to return.  At the moment, the only one I explicitly provide is `StringContent` that defaults to `application/json` as the contentType since that is by far the most common return value in use today, but the contentType is overridable.

If that feels weird, I can easily implement `JsonContent` as well, but I wasn't sure how I felt about that yet since that may step in our toes a bit if we ever decide to implement content negotiation since `JsonContent` is specifying the contentType directly.  We could also choose to deprecate it at that point but provide it in the short term to remove the weird feeling that the `StringContent` class is giving me.

WDYT?

**Comment by dcavanagh** [comment]
@dcherman I'm sorry about the delay. I have asked someone to review the changes.

**Comment by PodaruDragos** [comment]
@dcherman hey, sorry to bother you ( I know this is merged ) but I am currently in a middle of a big refactor. Upgrading typescript compiler options, packages etc.

`private apiController: BaseHttpController` is all the classes is never used. Was there a reason for this ?

I kinda want to remove those things, but iI want to know the reasoning for adding an unused parameter in every class

**Comment by dcherman** [comment]
@PodaruDragos Probably a case of YAGNI.

The original intent was to support creation of action results similar to https://github.com/aspnet/AspNetWebStack/blob/42991b3d2537b702736463f76a10a4fcf2ea44c9/src/System.Web.Http/Results/CreatedNegotiatedContentResult.cs#L48-L50, but obviously that never happened.

**Pull request merged**